src/toil/jobStores/aws/jobStore.py

```diff
@@ -63,30 +65,37 @@
                                       no_such_sdb_domain,
                                       sdb_unavailable,
                                       monkeyPatchSdbConnection,
-                                      retry_s3,
-                                      retryable_s3_errors,
                                       bucket_location_to_region,
-                                      region_to_bucket_location, copyKeyMultipart,
-                                      uploadFromPath, chunkedFileUpload, fileSizeAndTime)
+                                      region_to_bucket_location, fileSizeAndTime,
+                                      uploadFromPath, uploadFile, copyKeyMultipart)
+
 from toil.jobStores.utils import WritablePipe, ReadablePipe, ReadableTransformingPipe
 from toil.jobGraph import JobGraph
 import toil.lib.encryption as encryption
 
+
+def get_boto3_session():
+    botocore_session = botocore.session.get_session()
+    botocore_session.get_component('credential_provider').get_provider(
+        'assume-role').cache = botocore.credentials.JSONFileCache()
+    return boto3.Session(botocore_session=botocore_session)
+
+
 # Make sure to use credential caching when talking to Amazon via boto3
 # See https://github.com/boto/botocore/pull/1338/
-botocore_session = botocore.session.get_session()
-botocore_session.get_component('credential_provider').get_provider('assume-role').cache = botocore.credentials.JSONFileCache()
-boto3_session = boto3.Session(botocore_session=botocore_session)
+boto3_session = get_boto3_session()
```

Comment:
Why do we call `get_boto3_session()` again in some places, but use this module-level session in others?

Reply:
`get_boto3_session()` is called again in places that may need to be on its own thread (uploads and downloads). Since, according to the [boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html?highlight=multithreading#multithreading-and-multiprocessing),

> It is recommended to create a resource instance for each thread / process in a multithreaded or multiprocess application rather than sharing a single instance among the threads / processes.

and

> ... each thread would have its own Boto3 session and its own instance of the S3 resource.

The module-level session is only used by the module-level client, `s3_boto3_client`, which is only used in the static method `__getBucketRegion()` and when destroying a job store. This should be fine since low-level clients are thread safe.